### PR TITLE
feat: Add Ocient support

### PR DESCRIPTION
**Pull request by alexclavel-ocient**
**WIP**
### SUMMARY
Adding support for Ocient databases

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by jwilliams-ocient** [comment]
Thanks for putting this PR up @alexclavel-ocient !

A note for the rock star reviewers: expect 1 additional commit for the user doc 🙂 !

**Comment by betodealmeida** [comment]
This looks great! Feel free to ping me when when doc is up (here or on Slack) and I'll review it! <3

File: docs/docs/databases/ocient.mdx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+## Connecting to Ocient
+
+The connection string for Ocient looks like this:
+
+```shell
+ocient://{user}:{password}@{DNSname}:{port}/{database}
+```
+
+**NOTE**: You must enter the `user` and `password` credentials.  `host` defaults to localhost,
+port defaults to 4050, database defaults to `system` and `tls` defaults
```
Comment (betodealmeida) [inline docs/docs/databases/ocient.mdx:27]: How is `tls` passed to the URL?

File: superset/db_engine_specs/ocient.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+from superset import app
+from superset.models.core import Database
+from typing import Any, Callable, Dict, List, NamedTuple, Tuple, Optional, Pattern
+import threading
+
+from superset.models.sql_lab import Query
+
+# Ensure pyocient inherits Superset's logging level
+superset_log_level = app.config["LOG_LEVEL"]
+pyocient.logger.setLevel(superset_log_level)
```
Comment (betodealmeida) [inline superset/db_engine_specs/ocient.py:37]: Oh, this is clever! I'm going to steal this for some other DBs. 🙃

File: superset/db_engine_specs/ocient.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    engine_name = "Ocient"
+    # limit_method = LimitMethod.WRAP_SQL
+    force_column_alias_quotes = True
+    max_column_name_length = 30
+
+    # Store mapping of superset Query id -> Ocient ID
+    # These are inserted into the cache when executing the query
+    # They are then removed, either upon cancellation or query completion
+    query_id_mapping: Dict[str, str] = dict()
+    query_id_mapping_lock = threading.Lock()
```
Comment (betodealmeida) [inline superset/db_engine_specs/ocient.py:171]: Did you also test this with async queries? (with the query being executed by Celery)

**Review by betodealmeida — APPROVED** [review]
This looks really good, it's nice to see a DB engine spec so complete! 💪🏼 

I left a question about using `threading` and async query execution.

File: superset/db_engine_specs/ocient.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    def cancel_query(cls, cursor: Any, query: Query, cancel_query_id: str) -> bool:
+        with OcientEngineSpec.query_id_mapping_lock:
+            if query.id in OcientEngineSpec.query_id_mapping:
+                cursor.execute(f"CANCEL {OcientEngineSpec.query_id_mapping[query.id]}")
+                # Query has been cancelled, so we can safely remove the cursor from the cache
+                del OcientEngineSpec.query_id_mapping[query.id]
+
+                return True
+            # If the query is not in the cache, it must have either been cancelled elsewhere or completed
+            else:
```
Comment (betodealmeida) [inline superset/db_engine_specs/ocient.py:303]: The linter is going to complain about this `else:`, it's not needed.

File: docs/docs/databases/ocient.mdx (hunk 1 above)
Comment (jwilliams-ocient) [reply docs/docs/databases/ocient.mdx:27]: Great question! We should definitely provide an example here - will fix

**Comment by alexclavel-ocient** [comment]
@betodealmeida Any Idea why the CI cannot find pyocient? We have sqlalchemy-ocient in setup.py, and pyocient is a requirement for sqlalchemy-ocient.

**Comment by betodealmeida** [comment]
> @betodealmeida Any Idea why the CI cannot find pyocient? We have sqlalchemy-ocient in setup.py, and pyocient is a requirement for sqlalchemy-ocient.

The dependency in `setup.py` is optional, to have it run with CI you need to add it to `requirements/testing.in`:

https://github.com/apache/superset/blob/42db7e562d7d95ea57a030955e6bf68cca63fb23/requirements/testing.in#L19

File: superset/db_engine_specs/ocient.py (hunk 3 above)
Comment (jwilliams-ocient) [reply superset/db_engine_specs/ocient.py:173]: Hi @betodealmeida , I am the Ocient developer tasked with pushing this PR over the finish line, thanks for all of your help thus far!

Is there an integration test that stresses this async queries + ocient, or just a manual look-and-pray-for-racy-condition-if-exists? Pardon my ignorance, I found this [doc](https://superset.apache.org/docs/installation/async-queries-celery/), but couldn't find anything w.r.t. testing.

**Comment by jwilliams-ocient** [comment]
Is it possible for me to kick off a CI run using the Superset infra on my end?

**Comment by villebro** [comment]
Sure 👍

**Comment by jwilliams-ocient** [comment]
Ugh, once more, please? I'm an idiot and pushed a partial commit. 

In the event _this_ run fails causing me to have to push another commit,, should I just create a separate PR with a single, empty commit in order to gain write access / stop bugging your team for trivial hit-the-run-button tasks? @villebro 🙏🏾

**Comment by jwilliams-ocient** [comment]
Alright, @villebro , hopefully this is the last time I bug you today 😆. Here is a request for write access https://github.com/apache/superset/pull/23755

**Comment by villebro** [comment]
> Alright, @villebro , hopefully this is the last time I bug you today 😆. Here is a request for write access https://github.com/apache/superset/pull/23755

@jwilliams-ocient on a slightly related note, I recently read James Dyson's book "Invention, a life", in which he mentions that it took him 5127 prototypes before his first vacuum cleaner worked. So we're still far off from that figure 😂

**Comment by jwilliams-ocient** [comment]
Is the failure from `E2E / cypress-matrix (2, chrome) (pull_request) ` a flaky test? I'm unsure if the failure is related to our changes.

**Comment by jwilliams-ocient** [comment]
Hi @rusackas , @betodealmeida , and @villebro. On behalf of Ocient, I just wanted to say thank you for helping us push this PR through, much appreciated!

**Comment by rusackas** [comment]
Thanks to @jwilliams-ocient and @alexclavel-ocient for contributing this! If you want to talk about next steps to spread the word, hit me up on [Slack](http://bit.ly/join-superset-slack)!

**Pull request merged**